lxd/db/deployments.go

```diff
@@ -0,0 +1,1135 @@
+	}
+
+	defer func() { _ = insertDeploymentShapeConfigStmt.Close() }()
+
+	for deploymentShapeName, deploymentShape := range deploymentShapes {
+		// Marshal instance template.
+		deploymentShapeTemplateJSON, err := json.Marshal(deploymentShape.InstanceTemplate)
+		if err != nil {
+			return fmt.Errorf("Failed marshalling instance template: %w", err)
+		}
+
+		result, err := insertDeploymentShapeStmt.ExecContext(
+			ctx,
+			deploymentID,
+			deploymentShapeName,
+			deploymentShape.Description,
+			deploymentShape.ScalingMinimum,
+			deploymentShape.ScalingMaximum,
+			deploymentShape.ScalingCurrent,
+			string(deploymentShapeTemplateJSON),
+		)
+		if err != nil {
+			return fmt.Errorf("Failed inserting deployment shape: %w", err)
+		}
+
+		deploymentShapeID, err := result.LastInsertId()
+		if err != nil {
+			return err
+		}
+
+		for k, v := range deploymentShape.Config {
+			if v == "" {
+				continue
+			}
+
+			_, err = insertDeploymentShapeConfigStmt.ExecContext(ctx, deploymentShapeID, k, v)
+			if err != nil {
+				return fmt.Errorf("Failed inserting deployment shape config: %w", err)
+			}
+		}
+	}
+
+	return nil
+}
+
+// UpdateDeployment updates the deployment with the given ID.
+func (c *ClusterTx) UpdateDeployment(ctx context.Context, deploymentID int64, config *api.DeploymentPut) error {
+	_, err := c.tx.Exec(`
+			UPDATE deployments
+			SET description=?
+			WHERE id=?
+		`, config.Description, deploymentID)
+	if err != nil {
+		return err
+	}
+
+	_, err = c.tx.ExecContext(ctx, "DELETE FROM deployment_configs WHERE deployment_id=?", deploymentID)
+	if err != nil {
+		return err
+	}
+
+	err = deploymentConfigAdd(ctx, c.tx, deploymentID, config.Config)
+	if err != nil {
+		return err
+	}
+
+	_, err = c.tx.ExecContext(ctx, "DELETE FROM deployment_shapes WHERE deployment_id=?", deploymentID)
+	if err != nil {
+		return err
+	}
+
+	err = deploymentDeploymentShapesAdd(ctx, c.tx, deploymentID, config.DeploymentShapes)
+	if err != nil {
+		return err
+	}
+
+	return nil
+}
+
+// RenameDeployment renames a deployment.
+func (c *ClusterTx) RenameDeployment(ctx context.Context, deploymentID int64, newName string) error {
+	_, err := c.tx.ExecContext(ctx, "UPDATE deployments SET name=? WHERE id=?", newName, deploymentID)
+	return err
+}
+
+// DeleteDeployment deletes the deployment.
+func (c *ClusterTx) DeleteDeployment(ctx context.Context, deploymentID int64) error {
+	_, err := c.tx.ExecContext(ctx, "DELETE FROM deployments WHERE id=?", deploymentID)
+	return err
+}
+
+// GetDeploymentsURIs returns the URIs for the deployments with the given project.
+func (c *ClusterTx) GetDeploymentsURIs(ctx context.Context, projectID int, projectName string) ([]string, error) {
+	sql := `SELECT deployments.name from deployments WHERE deployments.project_id = ?`
+
+	names, err := query.SelectStrings(ctx, c.tx, sql, projectID)
+	if err != nil {
+		return nil, fmt.Errorf("Unable to get URIs for deployment: %w", err)
+	}
+
+	uris := make([]string, len(names))
```

Comment:
I tend to use `make(t,0,cap)`  and then use `append` in the for loop.
This avoids leaving any nil entries if the logic later changes to exclude some entries.
The current approach initialises all of entries to their zero value and then immediately overwrites them all (hopefully).